vernac/lemmas.ml

```diff
@@ -196,121 +189,92 @@ let retrieve_first_recthm uctx = function
-  let univs = UState.check_univ_decl ~poly uctx udecl in
-  let t_i = norm typ in
+let save_remaining_recthms ?fix_exn ~poly ~scope ~udecl ~hook uctx body opaque i { Recthm.name; typ; impargs } =
+  let sigma = Evd.from_ctx uctx in
   let kind = Decls.(IsAssumption Conjectural) in
   match body with
   | None ->
-    let open DeclareDef in
-    (match scope with
-     | Discharge ->
-       (* Let Fixpoint + Admitted gets turned into axiom so scope is Global,
-          see finish_admitted *)
-       assert false
-     | Global local ->
-       let kind = Decls.(IsAssumption Conjectural) in
-       let decl = Declare.ParameterEntry (None,(t_i,univs),None) in
-       let kn = Declare.declare_constant ~name ~local ~kind decl in
-       GlobRef.ConstRef kn, impargs)
+    let _, pe = DeclareDef.prepare_parameter ~allow_evars:false ~poly sigma udecl typ in
+    DeclareDef.declare_assumption ?fix_exn ~name ~scope ~hook ~impargs ~uctx pe
   | Some body ->
-    let body = norm body in
-    let rec body_i t = match Constr.kind t with
-      | Fix ((nv,0),decls) -> mkFix ((nv,i),decls)
-      | CoFix (0,decls) -> mkCoFix (i,decls)
-      | LetIn(na,t1,ty,t2) -> mkLetIn (na,t1,ty, body_i t2)
-      | Lambda(na,ty,t) -> mkLambda(na,ty,body_i t)
-      | App (t, args) -> mkApp (body_i t, args)
-      | _ ->
-        anomaly Pp.(str "Not a proof by induction: " ++ Printer.pr_constr_env env sigma body ++ str ".") in
-    let body_i = body_i body in
-    let open DeclareDef in
-    match scope with
-    | Discharge ->
-      let const = Declare.definition_entry ~types:t_i ~opaque:opaq ~univs body_i in
-      let c = Declare.SectionLocalDef const in
-      let () = Declare.declare_variable ~name ~kind c in
-      GlobRef.VarRef name, impargs
-    | Global local ->
-      let const = Declare.definition_entry ~types:t_i ~univs ~opaque:opaq body_i in
-      let kn = Declare.declare_constant ~name ~local ~kind (Declare.DefinitionEntry const) in
-      GlobRef.ConstRef kn, impargs
-
-(* This declares implicits and calls the hooks for all the theorems,
-   including the main one *)
-let process_recthms ?fix_exn ?hook env sigma uctx ~udecl ~poly ~scope dref imps other_thms =
-  let other_thms_data =
-    if List.is_empty other_thms then [] else
-      (* there are several theorems defined mutually *)
-      let body,opaq = retrieve_first_recthm uctx dref in
-      List.map_i (save_remaining_recthms env sigma ~poly ~scope ~udecl uctx body opaq) 1 other_thms in
-  let thms_data = (dref,imps)::other_thms_data in
-  List.iter (fun (dref,imps) ->
-      maybe_declare_manual_implicits false dref imps;
-      DeclareDef.Hook.(call ?fix_exn ?hook { S.uctx; obls = []; scope; dref})) thms_data
+    let body = EConstr.of_constr (select_body i body) in
+    (* XXX: we are normalizing twice here, entries do contain ground terms *)
+    let _, de = DeclareDef.prepare_definition ~allow_evars:false ~opaque ~poly sigma udecl ~types:(Some typ) ~body in
```

Comment:
Is this really the right way to work? It seems suspicious.

Reply:
Umm, no idea how to solve that right now; frankly I'm not sure I understand all the details of the way recthms are handled; note that this is not a regression in itself, IIANM we did normalize twice before too.

I guess we should just revisit the way we handle recthms in full; this seems to need a large refactoring tho; we would likely have to move the knowledge of that down `proof`. Would be worth it on the other hand, got any ideas?

[This is the kind of stuff we could really benefit from some whiteboard discussion IMO]

Reply:
Normalize twice is not too bad, but prepare_def also does minimize which I don't trust so much here, did we already do that twice?

Reply:
We did not; do you have any suggestion how to handle this?

I guess we should go ahead and refactor the mutual fixpoint code even more; I have an idea how to do that but it will be a bit invasive; basically I'll introduce a mutual_entry type that keeps the invariants, then it is projected.

Reply:
Yup, writing this cleanup, I like how it looks, should be ready in a few days.

Reply:
Cleanup done, have a look.